ocs_ci/deployment/metallb.py

```diff
@@ -0,0 +1,566 @@
+        metallb_inst_data = templating.load_yaml(constants.METALLB_INSTANCE_YAML)
+        metallb_inst_data.get("metadata").update({"namespace": self.namespace_lb})
+
+        metallb_inst_file = tempfile.NamedTemporaryFile(
+            mode="w+", prefix="metallb_instance", delete=False
+        )
+        templating.dump_data_to_temp_yaml(metallb_inst_data, metallb_inst_file.name)
+
+        exec_cmd(f"oc create -f {metallb_inst_file.name}", timeout=240)
+
+        return self.metallb_instance_created()
+
+    def create_ip_address_pool(self):
+        """
+        Create IP address pool for MetalLB
+        :return: True if IP address pool is created, False if creation failed and None if IP address pool already exists
+        """
+        reserved_ips_num = config.ENV_DATA.get("ips_to_reserve")
+        if not reserved_ips_num:
+            raise ValueError(
+                "Number of reserved IP addresses for MetalLB is not specified"
+            )
+        else:
+            for num in range(1, reserved_ips_num + 1):
+                self.hostnames.append(
+                    f"clustername-{config.ENV_DATA['cluster_name']}-{num}"
+                )
+
+        # common part for both platforms
+        ipaddresspool_data = templating.load_yaml(METALLB_IPADDRESSPOOL_PATH)
+        if self.namespace_lb != METALLB_DEFAULT_NAMESPACE:
+            ipaddresspool_data.get("metadata").update({"namespace": self.namespace_lb})
+
+        self.ip_address_pool_name = ipaddresspool_data.get("metadata").get("name")
+
+        if self.ip_address_pool_created():
+            logger.info(
+                f"IPAddressPool {self.ip_address_pool_name} already exists in the namespace {self.namespace_lb}"
+            )
+            return
+
+        if config.ENV_DATA["platform"] == constants.HCI_VSPHERE:
+
+            # due to circular import error, import is here
+            from ocs_ci.deployment.vmware import assign_ips
+
+            logger.info("Reserving IP addresses from IPAM and Creating IP address pool")
+            self.addresses_reserved = assign_ips(hosts=self.hostnames)
+
+            ip_addresses_with_mask = [ip + "/32" for ip in self.addresses_reserved]
+            ipaddresspool_data.get("spec").update({"addresses": ip_addresses_with_mask})
+
+        elif config.ENV_DATA["platform"] == constants.HCI_BAREMETAL:
+            cidr = config.ENV_DATA["machine_cidr"]
+            network = ipaddress.ip_network(cidr)
+            ip_list_by_cidr = list(network)
+            ip_list_for_hosted_clusters = list()
+
+            # skip ip addresses reserved for machines, Network, Gateway, and Broadcast; first 10
+            for i, ip in enumerate(ip_list_by_cidr):
+                if i < 10 + 1 or i == len(ip_list_by_cidr) - 1:
+                    continue
+                ip_list_for_hosted_clusters.append(f"{ip}/32")
+            ipaddresspool_data.get("spec").update(
+                {"addresses": ip_list_for_hosted_clusters}
+            )
```

Comment:
I would propose to make the list of available IPs for Bare Metal deployment explicitly configurable via configuration file. We have to have anyway specific configuration file for each Baremetal cluster, so specifying the IPs for the ipaddresspool will be just small update there and it will be more generic for various environments.

Reply:
I thought I am following some convention. Because I am skipping first IPs that are usually for Network, Gateway, Broadcast and machine IPS. 
This will leave us a variable number of IPs that are left by cidr and we need to know only CIDR for all the job.
 
I did this by several reasons: 

- to not expose ipaddresses everywhere in docs, chats, deployment job 
- save User(qe) from the burden to find correct IPs 
- avoid potential issues when IPs are selected incorrectly.

Reply:
> I thought I am following some convention. Because I am skipping first IPs that are usually for Network, Gateway, Broadcast and machine IPS. This will leave us a variable number of IPs that are left by cidr and we need to know only CIDR for all the job.

This is true for the two environment we have now, but it might not be for some other environment we will have in the future. For example the Primary IPs of the servers in IBM Cloud are assigned randomly from the subnet (not sequentially from the start.

> I did this by several reasons:
> 
>     * to not expose ipaddresses everywhere in docs, chats, deployment job

The "static" configuration for the particular environment will be part of the "credential" conf, so at least this part will be handled as other credentials. We will also have to make sure to mask it in the logs (if it is desirable).

>     * save User(qe) from the burden to find correct IPs

This will be one-time action for each BM environment - we have to prepare the configuration for all the other parameters for the cluster (configuration of each of the server and other network configurations), so this is just one more item in the existing config file for the particular environment.

>     * avoid potential issues when IPs are selected incorrectly.

That is the reason, why I would propose to set it statically once in configuration file.

Reply:
> This is true for the two environment we have now, but it might not be for some other environment we will have in the future.

The method create_ip_address_pool has if-else block to make an actions considering the platform. For now they are vSphere and IBM. Further we will prepare more and update the function if there will be deviation on IP's per setup protocol. 

Regarding to have a static IP list approach that designed to be universal for all the platforms. I am not sure we will have static IPs and no IP reservation functionality for all the new platforms. So in any case we most likely would need to alter the function `create_ip_address_pool()` for AWS or BareMetal with agent, or other new platforms.

---

If we still want to have files with all static IPs, can you pls prepare the configs, so I can see format of the static IP list on GitLab, if I understood you correctly. I will fetch and work with it within existing PR

Reply:
My note here was about the IBM BM platform, on vSphere the approach of reserving the IPs in IPAM server is correct.

About the format of the configuration file, it is basically up to you - what you will prefer/what will be easily re-usable in the code. For example something like this?
```yaml
ENV_DATA:
    metallb_ip_address_pool:
        - "192.0.2.12/32"
        - "192.0.2.13/32"
        - "192.0.2.16/28"
        ...
```

Reply:
I will update PR, could you please create this config files on gitlab repo. If I remember correctly I don't have permissions for MR.
The IPs are (if not changed):

baremetal1
subnet: 52.118.6.128/27
52.118.6.129  Gateway
52.118.6.130  provisioning VM
52.118.6.131  baremetal01
52.118.6.132  baremetal02
52.118.6.133  baremetal03
52.118.6.134  baremetal04
52.118.6.135  baremetal05
52.118.6.136  baremetal06
52.118.6.137  API
52.118.6.138  Ingress
52.118.6.139  - 52.118.6.158  metalLB ipaddress pool

baremetal2
subnet: 169.63.205.128/27
169.63.205.129  Gateway
169.63.205.130  provisioning VM
169.63.205.131  baremetal2-01
169.63.205.132  baremetal2-02
169.63.205.133  baremetal2-03
169.63.205.134  baremetal2-04
169.63.205.135  baremetal2-05
169.63.205.136  baremetal2-06
169.63.205.137  API
169.63.205.138  Ingress
169.63.205.139 - 169.63.205.158  metalLB ipaddress pool for baremetal2 cluster

Reply:
As discussed, I've added following configuration to the credentials file for each BM cluster:
```
ENV_DATA:
    ip_address_pool:
        - "..."
        - "..."
        ...
```

Reply:
added to the PR, all the logic to find IPs by using the cidr was removed.